### PR TITLE
Add new optional OIDCDisableUserInfo setting for OIDC auth provider

### LINKED ISSUE #19318: OIDC: add `resource` url-parameter for self-hosted ADFS
Hi there,

I encountered a problem when setting up OIDC auth provider over ADFS 2019, same as described [here](https://forum.proxmox.com/threads/openid-and-adfs-2019-server.110350/).

### Nomad version
v1.6.x

### Issue
There is no way to set up `resource` url-parameter for for self-hosted ADFS as OIDC provider.

### Proposals for implementation
It can be done via the optional parameter `OIDCResource` in the Nomad OIDC config and optional config method `WithResource` in [cap module](https://github.com/hashicorp/cap/blob/b9202f18d585bb1e82d6fcd260c18983ac4a1335/oidc/request.go#L209).

### PR DESCRIPTION
**Pull request by eumikhailov**
Add new optional OIDCDisableUserInfo setting for OIDC auth provider which disables a request to the identity provider to get OIDC UserInfo.

This options is helpful when your identity provider doesn't send any additional claims from the UserInfo endpoint, such as 
[Microsoft AD FS](https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/overview/ad-fs-faq#i-m-trying-to-get-more-claims-on-the-userinfo-endpoint--but-it-s-only-returning-subject--how-can-i-get-more-claims-) OIDC Provider:
> The AD FS UserInfo endpoint always returns the subject claim as specified in the OpenID standards. AD FS doesn't support additional claims requested via the UserInfo endpoint

Fixes #19318

### CODE REVIEW

**Comment by eumikhailov** [comment]
Hi @tgross!
Please take a look :)

**Comment by tgross** [comment]
@eumikhailov I've been away for the holidays and haven't had a chance to review this yet. But I'll need you to sign the CLA before I can do so.

**Comment by eumikhailov** [comment]
@tgross happy new year! 😃
PS I've signed CLA

File: website/content/api-docs/acl/auth-methods.mdx
Comment (tgross) [inline website/content/api-docs/acl/auth-methods.mdx:67]: This flag is preventing Nomad from making a request to the external provider of user info, right?  Let's explain that in a little more detail _why_ you'd want this. Also, booleans always default to `false`.

```suggestion
  - `OIDCDisableUserInfo` `(bool: false)` - When set to `true`, Nomad will not make a request to the identity provider to get OIDC UserInfo. You may wish to set this if your identity provider...TBD
```

**Review by tgross — COMMENTED** [review]
Hi @eumikhailov I've left a comment in https://github.com/hashicorp/nomad/issues/19318#issuecomment-1874500172 which we should probably discuss further before spending more time on this PR.

File: website/content/api-docs/acl/auth-methods.mdx
Comment (eumikhailov) [reply website/content/api-docs/acl/auth-methods.mdx:67]: Hi, @tgross!

I would suggest

> - `OIDCDisableUserInfo` `(bool: false)` - When set to `true`, Nomad will not make a request to the identity provider to get OIDC UserInfo. **You may wish to set this if your identity provider sending all necessary user claims in OIDC token**.

Should I add some links to MS FAQ?
> Microsoft OIDC provider returns only sub claim in UserInfo endpoint
And https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/overview/ad-fs-faq#i-m-trying-to-get-more-claims-on-the-userinfo-endpoint--but-it-s-only-returning-subject--how-can-i-get-more-claims-

Comment (tgross) [reply website/content/api-docs/acl/auth-methods.mdx]: I would not link to their FAQ, as once they re-org'ed MSDN you can't really rely on them keeping their links intact anymore.

**Review by tgross — COMMENTED** [review]
@eumikhailov I don't have an environment I can test this in. Have you verified that this actually works with ADFS?

**Comment by eumikhailov** [comment]
@tgross can I reuse nomad binary from CI builds?

File: website/content/api-docs/acl/auth-methods.mdx
Comment (eumikhailov) [reply website/content/api-docs/acl/auth-methods.mdx]: Ok, I've updated description comment for the history

**Comment by eumikhailov** [comment]
@tgross tested with Microsoft AD FS 4.0 and it's work!

1) Setup oidc
```bash
home/wsl/nomad# ./nomad acl auth-method info microsoft-oidc
Name              = microsoft-oidc
Type              = OIDC
Locality          = global
Max Token TTL     = 1h0m0s
Token Name Format = ${auth_method_type}-${auth_method_name}
Default           = false
Create Index      = 34
Modify Index      = 36

Auth Method Config

JWT Validation Public Keys = <none>
JWKS URL                   = <none>
OIDC Discovery URL         = ***/adfs
OIDC Client ID             = de2658c8-cfbc-4d65-9b5e-bc3e85d1552d
OIDC Client Secret         = <none>
OIDC Disable UserInfo      = true
OIDC Scopes                = https://nomad.***/openid,openid
Bound audiences            = de2658c8-cfbc-4d65-9b5e-bc3e85d1552d,https://nomad.***
Bound issuer               = <none>
Allowed redirects URIs     = http://localhost:4646/ui/settings/tokens
Discovery CA pem           = <none>
JWKS CA cert               = <none>
Signing algorithms         = <none>
Expiration Leeway          = 0s
NotBefore Leeway           = 0s
ClockSkew Leeway           = 0s
Claim mappings             = {upn: upn}
List claim mappings        = {roles: roles}
```

2) Setup binding rule
```
home/wsl/nomad# ./nomad acl binding-rule info 743327e9-117a-3041-7e1c-718df4516900
ID           = 743327e9-117a-3041-7e1c-718df4516900
Description  = <none>
Auth Method  = microsoft-oidc
Selector     = "\"role-mgmt\" in list.roles"
Bind Type    = management
Bind Name    = <none>
Create Time  = ***
Modify Time  = ***
Create Index = 42
Modify Index = 42
```

3) Logon in UI works fine
image

**Review by tgross — CHANGES_REQUESTED** [review]
Ok @eumikhailov I've tested this out as described in https://github.com/hashicorp/nomad/issues/19318#issuecomment-1883396357. I only need two more things from you:

* Can you rebase this on `main` to resolve the merge conflict?
* Can you run `make cl` to add a changelog entry?

Once that's done, I'll merge this and get it backported. Thanks!

**Comment by eumikhailov** [comment]
@tgross OK, fixed!

**Comment by eumikhailov** [comment]
@tgross OK, fixed!

**Review by tgross — APPROVED** [review]
LGTM! I'll merge once CI is green, and make sure the backports happen.

**Pull request merged**
